SB-1.md

```diff
@@ -0,0 +1,76 @@
+# SB-1 (AMQP Backed Implementation)
+
+**Status**: Under Review
+
+This proposal introduces a method of using a message queue (Anything implementing
+AMQP 0.9.1, focusing on RabbitMQ) rather than the internal, custom queue.
+
+## Context
+
+Much of Seabird core has been built around a custom message queue, limited to a
+single node. This has made more complex operations harder to implement and led
+to less robust code.
+
+## Goals
+
+- Remove custom internal message queue
+- Replace as much logic as possible with queue routing
+
+## Potential Problems
+
+- This loses information about the current streams, such as channels,
+  and backend info. This may be acceptable for now and is related to the API refactor in [SB-3](./SB-3.md)
+
+## Details
+
+There are two types of "requests" in Seabird: request-response and streaming.
+Both of these will need to be backed by a message queue. All of this will be
+handled by seabird-core, so the actual implementation of clients should not
+change.
+
+These request types refer to the internal implementation of seabird-core.
+
+Queues will come in three forms:
+
+- `seabird.reply.*`, a separate queue per RPC request, an exclusive, direct
+  queue. This could potentially have a name provided by rabbit.
+- `seabird.command.*`, a separate queue per backend connection, an exclusive,
+  direct queue. The usage will be described under request-response.
+- `seabird.events`, a single "firehose", header exchange. All backend events will
+  be submitted to this queue. On startup, seabird-core will attempt to create
+  this queue.
+
+### Request-Response
+
+On an incoming RPC request (as an example "SendMessageRequest") the following
+will happen in order:
+
+- seabird-core will create an exclusive, direct queue to be used for
+  the given RPC reply
```

Comment:
I'd love to see a bit more information here about AMQP. "Exclusive, direct queue" means something, but the barrier to entry of understanding is too high for this RFC I think.

Reply:
Yeah, that's fair.

Exclusive means only 1 producer and the queue is automatically removed when the creator exits. This can be used to tell when the message cannot be written to the queue (because the client making the request has closed, either by choice or timeout).

Direct is one of the types of exchanges:

- Direct - any message pushed to this queue will be delivered to one consumer. There is no other routing.
- Fanout - any message pushed to this queue will be delivered to all consumers. There is no other routing.
- Topic - any message pushed to this queue will be delivered to all matching consumers. Routing is based on "topics", which are dot-separated paths like "seabird.message" or "seabird.command".
- Header - any message pushed to this queue will be delivered to all matching consumers. Routing is based on "headers" which are key-value pairs. The matching can either match "all" or "any" of the headers, depending on what you need.